src/erc721/extensions/enumerable/module.cairo

```diff
@@ -0,0 +1,176 @@
+// Source: https://github.com/Seraph-Labs/Cairo-Contracts/blob/cairo-2/src/tokens/erc721/extensions/enumerable.cairo
+
+#[starknet::contract]
+mod ERC721Enumerable {
+    use option::OptionTrait;
+    use integer::BoundedInt;
+    use traits::{TryInto, Into};
+    use starknet::ContractAddress;
+    use cairo_erc_3525::src5::module::SRC5;
+    use cairo_erc_3525::erc721::module::ERC721;
+    use cairo_erc_3525::erc721::extensions::enumerable::interface::{IERC721_ENUMERABLE_ID, IERC721Enumerable};
+
+    #[storage]
+    struct Storage {
+        _supply: u256,
+        _index_to_tokens: LegacyMap::<u256, u256>,
+        _tokens_to_index: LegacyMap::<u256, u256>,
+        _owner_index_to_token: LegacyMap::<(ContractAddress, u256), u256>,
+        _owner_token_to_index: LegacyMap::<u256, u256>,
+    }
+
+    #[external(v0)]
+    impl ERC721EnumerableImpl of IERC721Enumerable<ContractState> {
+        fn total_supply(self: @ContractState) -> u256 {
+            self._supply.read()
+        }
+
+        fn token_by_index(self: @ContractState, index: u256) -> u256 {
+            // assert index is not out of bounds
+            let supply = self._supply.read();
+            assert(index < supply, 'ERC721Enum: index out of bounds');
+            self._index_to_tokens.read(index)
+        }
+
+        fn token_of_owner_by_index(
+            self: @ContractState, owner: ContractAddress, index: u256
+        ) -> u256 {
+            self._token_of_owner_by_index(owner, index).expect('ERC721Enum: index out of bounds')
+        }
+    }
+
+    #[generate_trait]
+    impl InternalImpl of InternalTrait {
+        fn initializer(ref self: ContractState) {
+            let mut unsafe_state = SRC5::unsafe_new_contract_state();
+            SRC5::InternalImpl::register_interface(
+                ref unsafe_state, IERC721_ENUMERABLE_ID
+            );
+        }
+
+        fn transfer_from(
+            ref self: ContractState, from: ContractAddress, to: ContractAddress, token_id: u256
+        ) {
+            self._remove_token_from_owner_enum(from, token_id);
+            self._add_token_to_owner_enum(to, token_id);
+
+            let mut unsafe_state = ERC721::unsafe_new_contract_state();
+            ERC721::ERC721Impl::transfer_from(ref unsafe_state, from, to, token_id);
+        }
+
+        fn safe_transfer_from(
```

Comment:
Check whether this method makes sense in the Starknet context (see [this discussion](https://github.com/starknet-io/SNIPs/discussions/36#discussioncomment-6640087))

Reply:
I'll keep it for now and we will let OZ make the standard implementation. So not a real problem I think